google/cloud/pubsub/internal/publisher_stub_factory_test.cc

```diff
@@ -0,0 +1,467 @@
+
+#include "google/cloud/pubsub/internal/publisher_stub_factory.h"
+#include "google/cloud/pubsub/testing/mock_publisher_stub.h"
+#include "google/cloud/credentials.h"
+#include "google/cloud/grpc_options.h"
+#include "google/cloud/internal/api_client_header.h"
+#include "google/cloud/testing_util/scoped_log.h"
+#include "google/cloud/testing_util/status_matchers.h"
+#include "google/cloud/testing_util/validate_metadata.h"
+#include <gmock/gmock.h>
+
+namespace google {
+namespace cloud {
+namespace pubsub_internal {
+GOOGLE_CLOUD_CPP_INLINE_NAMESPACE_BEGIN
+namespace {
+
+using ::google::cloud::pubsub_testing::MockPublisherStub;
+using ::google::cloud::testing_util::ScopedLog;
+using ::google::cloud::testing_util::StatusIs;
+using ::google::cloud::testing_util::ValidateMetadataFixture;
+using ::testing::Contains;
+using ::testing::HasSubstr;
+using ::testing::NotNull;
+
+using BasePublisherStubFactory = std::function<std::shared_ptr<PublisherStub>(
+    std::shared_ptr<grpc::Channel>)>;
+
+// The point of these tests is to verify that the `CreatePublisherStub` factory
+// function injects the right decorators. We do this by observing the
+// side effects of these decorators. All the tests have nearly identical
+// structure. They create a fully decorated stub, configured to round-robin
+// over kTestChannel mocks.  The first mock expects a call, the remaining mocks
+// expect no calls.  Some of these side effects can only be verified as part of
+// the first mock.
+
+class PublisherStubFactory : public ::testing::Test {
+ protected:
+  void IsContextMDValid(grpc::ClientContext& context, std::string const& method,
+                        google::protobuf::Message const& request) {
+    return validate_metadata_fixture_.IsContextMDValid(
+        context, method, request,
+        google::cloud::internal::HandCraftedLibClientHeader());
+  }
+
+ private:
+  ValidateMetadataFixture validate_metadata_fixture_;
+};
+
+auto constexpr kTestChannels = 3;
+using MockFactory = ::testing::MockFunction<std::shared_ptr<PublisherStub>(
+    std::shared_ptr<grpc::Channel>)>;
+static_assert(std::is_same<decltype(MockFactory{}.AsStdFunction()),
+                           BasePublisherStubFactory>::value,
+              "Mismatched mock factory type");
+std::shared_ptr<PublisherStub> CreateTestStub(
+    CompletionQueue cq, BasePublisherStubFactory const& factory) {
+  auto credentials = google::cloud::MakeAccessTokenCredentials(
+      "test-only-invalid",
+      std::chrono::system_clock::now() + std::chrono::minutes(5));
+  return CreateDecoratedStubs(std::move(cq),
+                              google::cloud::Options{}
+                                  .set<GrpcNumChannelsOption>(kTestChannels)
+                                  .set<TracingComponentsOption>({"rpc"})
+                                  .set<UnifiedCredentialsOption>(credentials),
+                              factory);
+}
+
+TEST_F(PublisherStubFactory, CreateTopic) {
```

Comment:
FYI: I think these days, I would have one test for:
- RoundRobin
- Auth
- Metadata
- Logging
- Tracing (eventually)

and just use the `CreateTopic` RPC for all of them... but you already wrote the tests so we should just leave them as is.

Reply:
When you add the tracing stub, I'd just test it on 1 RPC.

Reply:
Ack, added the separate unit tests anyway and the tracing one can go with them